pkg/operator/telemetry/telemetry.go

```diff
@@ -0,0 +1,242 @@
+// Package telemetry provides telemetry functionality for the ToolHive operator.
+package telemetry
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/stacklok/toolhive/pkg/updates"
+	"github.com/stacklok/toolhive/pkg/versions"
+)
+
+const (
+	// updateInterval defines how often to check for updates
+	updateInterval = 4 * time.Hour
+	// configMapName is the name of the ConfigMap used to store telemetry data
+	configMapName = "toolhive-operator-telemetry"
+	// configMapNamespace is the namespace where the ConfigMap is stored
+	configMapNamespace = "toolhive-system"
+	// instanceIDKey is the key used to store the instance ID in the ConfigMap
+	instanceIDKey = "instance-id"
+)
+
+// Service handles telemetry operations for the operator
+type Service struct {
+	client        client.Client
+	versionClient updates.VersionClient
+	namespace     string
+}
+
+// LeaderTelemetryRunnable runs telemetry checks only when this instance is the leader
+type LeaderTelemetryRunnable struct {
+	TelemetryService *Service
+}
+
+// Start starts the telemetry runner
+func (t *LeaderTelemetryRunnable) Start(ctx context.Context) error {
+	ctxLogger := log.FromContext(ctx)
+	ctxLogger.Info("Leader elected, starting telemetry worker")
+
+	// Start telemetry worker in a goroutine with the leader context
+	// When leadership is lost, ctx will be cancelled and telemetry will stop
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				ctxLogger.Error(fmt.Errorf("telemetry worker panic: %v", r), "Telemetry worker panicked")
+			}
+		}()
+		t.TelemetryService.StartTelemetryWorker(ctx)
+	}()
+
+	// Wait for context cancellation (leadership lost or shutdown)
+	<-ctx.Done()
+	ctxLogger.Info("Leadership lost, telemetry worker stopped")
+	return nil
+}
+
+// NeedsLeaderElection indicates whether this runnable needs leader election
+func (*LeaderTelemetryRunnable) NeedsLeaderElection() bool {
```

Comment:
this is unused?

Reply:
https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#Server.NeedLeaderElection